Initial script edit

caelanradford opened the pull request:
This adds the initial edited script from CodonTilingPrimers and an updated README to describe what the changes do. Here is a short summary: 
- You must now pass a csv file as an argument that specifies what mutants to make at what sites 
- You must now pass a codon frequency csv file as an argument that determines what codon will to mutated to for each amino acid 
- The gene sequence given in a text file should now be uppercase for the WHOLE gene, including the start and stop codons 
- The gene sequence given in a text file no longer necessarily has to have lower case sequence before and after the gene; there must only be enough room for primer overhangs before the first codon mutagenized and after the last codon mutagenized
- As is written right now, there is no option for degenerate codons, which is why this is a new repository. This functionality could be added back later 

More functionality could be added later, like randomly picking codons for an amino acid as an option, or picking weighted on frequencies, etc. If this edit looks ok, I'll make a test case with BF520 primers and add them to the repo. I am still also making sure it is working properly; just wanted to make sure this approach looked ok in the meantime.

caelanradford committed:
Initial edit of create_primers.py and README

caelanradford committed:
Adds missing argument to create_primers.py

I accidentally left out the description and a few references to codon_frequency_csv in the script.

jbloom commented:
@khdusenbury or @fwelsh, are one or both of you able to review this?

khdcrawford commented:
I am happy to review, but probably tomorrow (or tonight at the earliest).

fwelsh commented:
I can look over it this afternoon as well

khdcrawford commented on README.md:
```diff
@@ -1,2 +1,58 @@
+Codon mutagenesis can be used to create mutant libraries of a gene with all possible codon mutations.
+Such libraries are useful for experiments such as [deep mutational scanning](https://www.ncbi.nlm.nih.gov/pubmed/25075907) or [mutational antigenic profiling](http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271)
+
+A [script in a separate repository](https://github.com/jbloomlab/CodonTilingPrimers) can be used to generate random mutations at each codon in a gene. This protocol randomly mutates each codon to all other possible codons. This protocol and papers using it are also described in that repository.
```
I _think_ the CodonTilingPrimers script can mutate all codons to any specified other codon? That's not necessary to mention, but I would mention that it's using degenerate nucleotides as a contrast to the new script, so maybe say "This protocol randomly mutates each codon to all other possible amino acids using degenerate nucleotides such as `NNS`" 

Because also with `NNS`, we're not mutating to all other codons, but to all other amino acids.

caelanradford replied:
I tried to clarify this more

khdcrawford commented on README.md:
```diff
@@ -1,2 +1,58 @@
+
+A [script in a separate repository](https://github.com/jbloomlab/CodonTilingPrimers) can be used to generate random mutations at each codon in a gene. This protocol randomly mutates each codon to all other possible codons. This protocol and papers using it are also described in that repository.
+
+Rather than generating mutant libraries with all possible codon mutations at every site in a gene, the script in this repository can be used to generate primers that will result in specific desired mutations at each site in a gene. In this way, a mutant library can be generated which has fewer lethal mutations than random codon mutagenesis.
```
Probably say "all possible amino acid mutations" since we aren't necessarily doing al codons anymore.

khdcrawford commented on README.md:
```diff
@@ -1,2 +1,58 @@
+
+*seq* : Text file containing the sequence of the gene. The gene itself should be upper case, including the start and stop codons. Any flanking regions should be lower case. There must be >= (primerlength - 3) / 2.0 nucleotides before the first codon mutagenized and after the last codon mutagenized, otherwise there would not be enough flanking nucleotides for a primer to mutate those codons.
+
+*mutations_csv* : csv file containing a table with the mutations the primers will make to seq. The table should have a column, 'site', with site numbers, and a column, 'mutant', with the single letter amino acid mutant to make at that site. Each site can have multiple mutants, or none. The site number should denote the number of the codon in the uppercase gene, with the start codon being 1.
```
If a site has multiple mutants, does it need to be listed multiple times, or can you supply a list to the 'mutant' column?

caelanradford replied:
I chose to have each mutation be a line in the file so that it would be possible later to add the ability to specify how many codons to generate for that mutation or how to choose the codons (which would take adding another column).

khdcrawford replied:
Sounds good, you should just make it clear then that if a site has multiple mutants, it needs to be listed separately for each mutant. That's not clear from the description.

khdcrawford commented on README.md:
```diff
@@ -1,2 +1,58 @@
+
+*mutations_csv* : csv file containing a table with the mutations the primers will make to seq. The table should have a column, 'site', with site numbers, and a column, 'mutant', with the single letter amino acid mutant to make at that site. Each site can have multiple mutants, or none. The site number should denote the number of the codon in the uppercase gene, with the start codon being 1.
+
+*codon_frequency_csv* : csv file containing a table with codon frequencies used to determine which codons sites will be mutated to. The table should have the columns 'aa', 'codon', and 'frequency'. Each row should have one single letter amino acid in 'aa', one codon corresponding to that amino acid in 'codon', and the frequency of that codon in 'frequency'. The script uses the codon with the high ...
```
Haven't look at code yet, so not sure if it's worth changing now or how it would work, but I'm pretty sure you can just import codon frequency tables (cai tables I think they're called, but I'm not sure what that stands for) using BioPython.

caelanradford replied:
I think BioPython can be used to generate codon frequency tables or use them to score sequences on their codon usage, but I did not find any well documented functionality to back translate using them. If you can find something like that I would use it.

khdcrawford replied:
I need to think more about the back translation, but why exactly is it necessary? Aren't you specifying mutants by aa, so why do you need to back translate from codon?

caelanradford replied:
If you specify what mutants you want by aa, then unless you want to make primers for each codon for that aa, you have to pick a codon (back translate the aa to a codon). The codon frequency table is used so that the codon picked isn't just random, but is the most frequent codon for that aa in whatever organism you supply the table for. I am planning on adding an option for just random, or semi random like you suggested.

khdcrawford replied:
Gotcha. I thought that we had used a built in function for this in `serologyseq` (https://github.com/jbloomlab/serologyseq), but it looks like we did write our own dictionary to map amino acids and codons. We did just use built in cai tables though, so there might be something helpful in the `design_oligos.py` script (https://github.com/jbloomlab/serologyseq/blob/master/src/design_oligos.py) if you didn't want to have to have the user input their own codon_frequency table, but it's not as easy as I thought it was, so what you have with the user inputting their own frequency table is also fine.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+
+Primers containing specified mutageneic codons at their center are tiled along a gene.
+
+Edited by Caelan Radford in 2020, from the CodonTilingPrimers code written by Jesse Bloom in 2013 and edited by Adam Dingens Nov 2015 to generate primers of differing lengths to all have a Tm of ~60C.
```
Could you add new lines to make this doc string easier to read? In python, lines should technically be no more than 79 characters long.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+            formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+            )
+
+    parser.add_argument('sequencefile', help="the name of a file giving the sequence for which we are designing the primers. This file should only contain the sequence, and should not have any headers or other content. For the sequence, make the 5' and 3' ends that you do not want to mutate in lower case. Make the portion of the coding sequence that you want to tile with ambiguous codons in upper ...
```
Also please continue these strings onto new lines to maintain line length <= 79 characters.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+
+def ReverseComplement(seq):
+    """Returns reverse complement of sequence. Preserves upper/lower case."""
+    d = {'A':'T', 'T':'A', 'C':'G', 'G':'C', 'a':'t', 't':'a', 'c':'g', 'g':'c', 'N':'N', 'n':'n', 'S':'S', 's':'s', 'K':'M', 'k':'m'}
```
Is there a reason my `S`, `K`, and `M` are the only ambiguous nucleotides included?

caelanradford replied:
This was a copy over from the old script. We could add more, which ones do you want to add?

khdcrawford replied:
I guess my main point is that this should be deliberate and not just a copy/paste, so it should include what you think is necessary for the script (or maybe all of the degenerate nucleotides if you think at one point this script might allow degenerate nucleotide functionality).

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    n = len(seq)
+    assert primerlength % 2 == 1, "primer length not odd"
+    initial_flanklength = (primerlength - 3) // 2
+    upperseq = ''.join([nt for nt in seq if nt.istitle()])
```
I think this should be using `nt.isupper()`. Technically `istitle()` works, but you're just comparing one nt at a time, so `isupper()` seems better.

You can also probably use `isupper()` on the seq somehow and not have to iterate to get all upper case letters back out, but I haven't confirmed the syntax for that.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    initial_flanklength = (primerlength - 3) // 2
+    upperseq = ''.join([nt for nt in seq if nt.istitle()])
+    assert upperseq in seq, "upper case nucleotides not substring"
+    assert len(upperseq) % 3 == 0, "length of upper case not multiple of 3"
```
This is a small thing, but I think @jbloom told me recently that `assert` statements are good for debugging, but for actual error checking that you want to be part of your final code, using `raise ValueError` is better. @jbloom can clarify perhaps?

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+
+    Returns a list of all these primers as *(name, sequence)* 2-tuples.
+    """
+    n = len(seq)
```
Do you use `n` anywhere?

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    try:
+        sites = df['site'].tolist()
+    except:
+        print('mutations_csv must have column "site"')
```
This is a small and potentially just a preference thing, but if you're going to have a string with both ' and ", I think it's best to have the " on the outside and the ' specifying the inside quotes. Like: "`mutations_csv` must have column 'site'"

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+        else:
+            mutations_dict[site].append(mutations[i])
+    # Make sure this is enough flanking sequence to make primers
+    initial_flanklength = (primerlength - 3) // 2
```
You already set this variable in line 93. Just set it in one place.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    startupper = seq.index(upperseq) + ((firstcodon - 1) * 3)
+    if startupper < initial_flanklength:
+        raise ValueError("not enough 5' flanking nucleotides")
+    if n - len(upperseq) - seq.index(upperseq) + (len(upperseq) - (lastcodon - 1) * 3) < initial_flanklength:
```
Oh here is where you use `n`. Maybe just define `n` closer to where you use it.

khdcrawford replied:
This also took me a long time to parse, maybe provide a comment above it saying what each part is measuring.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    if n - len(upperseq) - seq.index(upperseq) + (len(upperseq) - (lastcodon - 1) * 3) < initial_flanklength:
+        raise ValueError("not enough 3' flanking nucleotides")
+    # Read in the codon frequency table, make dict for back translating
+    df = pd.read_csv('codon_frequencies.csv')
```
I'm not sure, but I think this is maybe where you could use the built in BioPython cai tables?

I _think_ that functionality might allow you to go from aa -> codon and back more easily.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+        # Replace ambiguous_codon with codon insert later
+        codon_inserts = []
+        if icodon + 1 in mutations_dict:
+            for mutation in mutations_dict[icodon + 1]:
```
Probably not necessary now, but in terms of extensions of this code, you could imagine adding a `--codon_choice` flag and have it be either `max` or `random` to specify if the codon chosen is always the one with the max frequency or if you choose randomly (maybe for random choice, you'd want to set a threshold minimum frequency needed for a codon to be included in the `back_t_dict`)

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+    for icodon in range(ncodons):
+        # Determine what the insert should be for the codon
+        # Replace ambiguous_codon with codon insert later
+        codon_inserts = []
```
Do you do anything with this list?

caelanradford replied:
It was from an earlier version that I forgot to delete, thanks.

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+        raise IOError(f"Cannot find mutations_csv {mutations_csv}")
+
+    # Design forward mutation primers
+    mutforprimers = CreateMutForOligosVarLength(sequence, mutations_csv, primerlength, primerprefix, args['maxprimertm'], args['minprimertm'], args['maxlength'], args['minlength'])
```
Please break onto multiple lines!

khdcrawford commented on create_primers.py:
```diff
@@ -0,0 +1,261 @@
+                max_frequency = max(back_t_dict[mutation])
+                codon_insert = back_t_dict[mutation][max_frequency]
+                i = startupper + icodon * 3
+                primer = "%s%s%s" % (seq[i - initial_flanklength : i], codon_insert, seq[i + 3 : i + 3 + initial_flanklength])
```
You use this in a lot of places, but I'm just commenting here. The "%s", "%d", etc. string formatting is a hold over from the earlier script, but it is also very Python 2. 

I'd recommend switching all your strings to f string formatting, which is much more readable in my opinion too. So the code would be:

`primer = f"{seq[i - initial_flanklength : i]}{codon_insert}{seq[i + 3 : i + 3 + initial_flanklength]}"`

Like in the line below.

I know there are lots of places where this change would need to be made, but I think it's worth making in order to be consistent and up-to-date.

khdcrawford reviewed (changes requested):
Overall, this looks good Caelan. I there are a couple things to clean up, but most of it is just formatting. I would definitely add some tests and I do think the functionality of picking max vs. random codon would be good, probably not necessary, so I'd say only do it if you want.

caelanradford committed:
Adds suggested edits

I made a lot of Kate's suggested edits. I got rid of the old string syntax, and changed them to f strings. I made the arguments comments more readable. I added a few lines of code in different places to break things up for the sake of readability. Also slightly changed some wording in the README.

caelanradford commented:
I made a lot of Kate's suggested edits. I got rid of the old string syntax, and changed them to f strings. I made the arguments comments more readable. I added a few lines of code in different places to break things up for the sake of readability. Also slightly changed some wording in the README. I also fixed a bug that I had missed before that was making the first codon primers were generated before too late in the gene. 

I am planning on adding some of the options Kate mentioned. I am planning on adding the ability to specify whether you want codons chosen based on frequencies or chosen at random, or some combination option (random pick from frequencies above some threshold). I am also planning on adding an option to choose how many codons you want for each mutation. One of the reasons I chose to have the mutations passed in a csv with a row for each mutation rather than each site is so that you could potentially change the number of codons you are targeting by mutation rather than just by site.

khdcrawford commented on create_primers.py:
```diff
@@ -32,33 +46,63 @@
-    parser.add_argument('codon_frequency_csv', help="the name of a csv file with a table of codon frequencies to be used to determine which codons to mutate to. It should have one column 'aa' with single letter amino acids, one column 'codon' with codons, and one column 'frequency' with codon frequencies. The highest frequency codon for each amino acid is used.")
-    parser.add_argument('primerprefix', help="prefix name to be added to each primer")
+    parser.add_argument('sequencefile', help="the name of a file giving the "
+    "sequence for which we are designing the primers. This file should only "
```
Sorry for more knit-picky style edits, but these continuation lines should be indented either two tabs or equal with the original open parenthesis. 

I suggest running `flake8` on your code as it will catch these style things and also give somewhat helpful error messages as to how to fix the style.

khdcrawford commented on create_primers.py:
```diff
@@ -32,33 +46,63 @@
-def CreateMutForOligosVarLength(seq, mutations_csv, codon_frequency_csv, primerlength, prefix, maxprimertm, minprimertm, maxlength, minlength):
-    """Creates oligos to tile a gene and introduce specified amino acid mutations at each site.
+def CreateMutForOligosVarLength(seq, mutations_csv, codon_frequency_csv,
+primerlength, prefix, maxprimertm, minprimertm, maxlength, minlength):
```
Same indent thing here regarding the continuation line (I think it's best to be indented equal to the open parenthesis, but I think 2 tabs is technically also acceptable, but I don't quite remember).

khdcrawford commented on create_primers.py:
```diff
@@ -146,59 +194,64 @@ def CreateMutForOligosVarLength(seq, mutations_csv, codon_frequency_csv, primerl
                 primerseq = Seq(primer)
 
-                TmNN = ('%0.2f' % mt.Tm_NN(primerseq, strict=False))
+                TmNN = mt.Tm_NN(primerseq, strict=False)
```
I'm not sure if this will cause issues or not, but you're now no longer rounding the Tm and it's not a string anymore. 

I haven't dug into the code enough to know if this is an issue, but worth keeping in mind/maybe writing some tests to confirm that's okay?

caelanradford replied:
I checked when playing around with the code, and it doesn't seem to do anything when no longer rounding it. It does make it easier to look at primer temps if you decide to print them out to check they are coming out in the range you want, so maybe I'll revert the change.

khdcrawford commented on create_primers.py:
```diff
@@ -146,59 +194,64 @@ def CreateMutForOligosVarLength(seq, mutations_csv, codon_frequency_csv, primerl
                 add_3 = True
                 minus_5 = True
                 flank5 = flank3 = initial_flanklength
                 if float(TmNN) > float(maxprimertm):
```
TmNN is now no longer a string, so I'm not sure if it's necessary to have `float()` around it, but I'd check all of this and make sure it doesn't mess up things anywhere that you're not rounding `TmNN` anymore. I could see some benefit to strictly defining how many digits `TmNN` includes, which you're not doing anymore.

khdcrawford commented on create_primers.py:
```diff
@@ -240,32 +293,25 @@ def main():
-        f.write("\r\nPlate %d\r\n" % iplate)
-        n_in_plate = 0
         for (name, primer) in primers:
-            f.write("%s, %s\r\n" % (name, primer))
```
Do you need the plate functionality? Or maybe it should be an argument for "order in plate" or not? I'm not sure how you order these things and if just a long list is sufficient, but I assume Adam had this functionality for a reason? But that may also be out-of-date.

caelanradford replied:
The in-plate ordering functionality was for ordering from IDT in plates and then hand mixing the primers. It seems the best options now are to order oligo pools pre-pooled, which we just need a list of the names and primers for.

khdcrawford replied:
Oh yeah, that sounds way better! Sweet.

khdcrawford reviewed (approved):
Looks great @caelanradford. I had a couple more syntax things, and then I think you should run `flake8` on this and make sure it passes before you merge, but I'm going to approve so you can merge when you're ready.

caelanradford committed:
Flake8 suggested changes and revert Tm rounding

I decided to revert the change of the Tm rounding. It does not seem to affect the code if it is changed but it could be useful to have the primer temperatures be readable for later debugging or output.
